articles/sliding-window-maximum.md

```diff
@@ -0,0 +1,899 @@
+        }
+
+        return output;
+    }
+}
+```
+
+```csharp
+public class Solution {
+    public int[] MaxSlidingWindow(int[] nums, int k) {
+        int n = nums.Length;
+        int[] leftMax = new int[n];
+        int[] rightMax = new int[n];
+
+        leftMax[0] = nums[0];
+        rightMax[n - 1] = nums[n - 1];
+
+        for (int i = 1; i < n; i++) {
+            if (i % k == 0) {
+                leftMax[i] = nums[i];
+            } else {
+                leftMax[i] = Math.Max(leftMax[i - 1], nums[i]);
+            }
+
+            if ((n - 1 - i) % k == 0) {
+                rightMax[n - 1 - i] = nums[n - 1 - i];
+            } else {
+                rightMax[n - 1 - i] = Math.Max(rightMax[n - i], nums[n - 1 - i]);
+            }
+        }
+
+        int[] output = new int[n - k + 1];
+
+        for (int i = 0; i < n - k + 1; i++) {
+            output[i] = Math.Max(leftMax[i + k - 1], rightMax[i]);
+        }
+
+        return output;
+    }
+}
+```
+
+::tabs-end
+
+### Time & Space Complexity
+
+* Time complexity: $O(n)$
+* Space complexity: $O(n)$
+
+---
+
+## 5. Deque
+
+::tabs-start
+
+```python
+class Solution:
+    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
+        output = []
+        q = deque()  # index
+        l = r = 0
+
+        while r < len(nums):
+            while q and nums[q[-1]] < nums[r]:
+                q.pop()
+            q.append(r)
+
+            if l > q[0]:
+                q.popleft()
+
+            if (r + 1) >= k:
+                output.append(nums[q[0]])
+                l += 1
+            r += 1
+
+        return output
+```
+
+```java
+public class Solution {
+    public int[] maxSlidingWindow(int[] nums, int k) {
+        int n = nums.length;
+        int[] output = new int[n - k + 1];
+        Deque<Integer> q = new LinkedList<>();
+        int l = 0, r = 0;
+
+        while (r < n) {
+            while (!q.isEmpty() && nums[q.getLast()] < nums[r]) {
+                q.removeLast();
+            }
+            q.addLast(r);
+
+            if (l > q.getFirst()) {
+                q.removeFirst();
+            }
+
+            if ((r + 1) >= k) {
+                output[l] = nums[q.getFirst()];
+                l++;
+            }
+            r++;
+        }
+
+        return output;
+    }
+}
+```
+
+```cpp
+class Solution {
+public:
+    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
+        int n = nums.size();
+        vector<int> output(n - k + 1);
+        deque<int> q;
+        int l = 0, r = 0;
+
+        while (r < n) {
+            while (!q.empty() && nums[q.back()] < nums[r]) {
+                q.pop_back();
+            }
+            q.push_back(r);
+
+            if (l > q.front()) {
+                q.pop_front();
+            }
+
+            if ((r + 1) >= k) {
+                output[l] = nums[q.front()];
+                l++;
+            }
+            r++;
+        }
+
+        return output;
+    }
+};
+```
+
+```javascript
```

Comment:
Good catch. I will add the deque class to the environment shortly. I made some minor corrections to your JS deque solution. Please update it and then I will merge this PR:

---

class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    maxSlidingWindow(nums, k) {
        const n = nums.length;
        const output = new Array(n - k + 1);
        const q = new Deque();
        let l = 0, r = 0;

        while (r < n) {
            while (q.size() && nums[q.back()] < nums[r]) {
                q.popBack();
            }
            q.pushBack(r);

            if (l > q.front()) {
                q.popFront();
            }

            if ((r + 1) >= k) {
                output[l] = nums[q.front()];
                l++;
            }
            r++;
        }

        return output;
    }
}